reconcile/external_resources/secrets_sync.py

```diff
@@ -159,56 +165,77 @@ def __init__(
+        self.vault_client = vault_client
+        self.vault_path = vault_path
 
     def _get_spec_hash(self, spec: ExternalResourceSpec) -> str:
         secret_key = f"{spec.provision_provider}-{spec.provisioner_name}-{spec.provider}-{spec.identifier}"
         return shake_128(secret_key.encode("utf-8")).hexdigest(16)
 
+    def _get_spec_outputs_secret_name(self, spec: ExternalResourceSpec) -> str:
+        return "external-resources-output-" + self._get_spec_hash(spec)
+
     def _populate_secret_data(self, specs: Iterable[ExternalResourceSpec]) -> None:
         if not specs:
             return
-        secrets_map = {
-            "external-resources-output-" + self._get_spec_hash(spec): spec
-            for spec in specs
-        }
+
+        secrets_map = {self._get_spec_outputs_secret_name(spec): spec for spec in specs}
+
         secrets = self.oc.get_items(
             "Secret", namespace=self.namespace, resource_names=list(secrets_map.keys())
         )
+
         for secret in secrets:
             secret_name = secret["metadata"]["name"]
             spec = secrets_map[secret_name]
             data = dict[str, str]()
             for k, v in secret["data"].items():
                 decoded = base64.b64decode(v).decode("utf-8")
+
                 if decoded.startswith("__vault__:"):
                     _secret_ref = json.loads(decoded.replace("__vault__:", ""))
                     secret_ref = VaultSecret(**_secret_ref)
                     data[k] = self.secrets_reader.read_secret(secret_ref)
                 else:
                     data[k] = decoded
-            spec.secret = data
 
-        self.source_secrets = list(secrets_map.keys())
+            spec.secret = data
 
-    def _delete_source_secrets(self) -> None:
-        for secret_name in self.source_secrets:
-            print("Deleting secret " + secret_name)
+    def _delete_source_secrets(self, specs: Iterable[ExternalResourceSpec]) -> None:
+        for spec in [s for s in specs if s.secret]:
+            secret_name = "external-resources-output-" + self._get_spec_hash(spec)
+            logging.debug("Deleting secret " + secret_name)
             self.oc.delete(namespace=self.namespace, kind="Secret", name=secret_name)
 
+    def _write_secrets_to_vault(self, specs: Iterable[ExternalResourceSpec]) -> None:
+        for spec in [s for s in specs if s.secret]:
+            secret_path = f"{self.vault_path}/{spec.cluster_name}/{spec.namespace_name}/{spec.identifier}"
+            stringified_secret = {k: str(v) for k, v in spec.secret.items()}
+            desired_secret = {"path": secret_path, "data": stringified_secret}
+            self.vault_client.write(desired_secret, decode_base64=False)  # type: ignore[attr-defined]
+
     def sync_secrets(self, specs: Iterable[ExternalResourceSpec]) -> None:
         super().sync_secrets(specs)
-        self._delete_source_secrets()
+        self._write_secrets_to_vault(specs)
```

Comment:
this call is new, is it missed in previous logic?

Reply:
Vault interaction was put off until now, this PR introduces all vault interactions.